R/utils.R

```diff
@@ -1,6 +1,8 @@
 
 read_char <- function(path, ...) {
-  readChar(path, nchars = file.info(path)$size, ...)
+  enc2utf8(
+    readChar(path, nchars = file.info(path)$size, useBytes = TRUE, ...)
```

Comment:
This does not seem good to me. E.g are you sure that the output of the check is in UTF-8 on Windows?

Reply:
Actually, this code does not assume UTF-8 output, so forget that question. But can you explain why this is good? It seems to me that `useBytes = TRUE` will just mark the output as `bytes`. Is that what we want?

Reply:
> It seems to me that useBytes = TRUE will just mark the output as bytes.

I don't think that's right (the result is still a character string, not `bytes`). From my understanding, `useBytes = TRUE` is generally a good idea since it [prevents R from translating to native encoding](https://kevinushey.github.io/blog/2018/02/21/string-encoding-and-r/), but I'm not sure that's actually true for `readChar()`. What I do know for sure is that `useBytes = TRUE` fixes the 1st error in #150, and other RStudio packages (e.g., shiny, shinytest, htmltools, etc) use it extensively.

Also, interestingly, I can read in the `Rout.fail` file with `brio::read_file()`, but it still errors out on `nchar()`

```r
path <- "ipc-testthat.Rout.fail"
txt <- brio::read_file(path)
Encoding(txt)
#> [1] "UTF-8"
nchar(txt)
#> Error in nchar(txt) : invalid multibyte string, element 1
```

Here's more info about what's in the file. Perhaps the 'correct' fix here is for the `Rout.fail` files to be encoded as UTF-8 when they're written to disk, but perhaps that's a bug in R that we should try to workaround?

```r
path <- "ipc-testthat.Rout.fail"
txt <- xfun::read_utf8(path)
#> Warning message:
#> In xfun::read_utf8(path) :
#>  The file ipc-testthat.Rout.fail is not encoded in UTF-8. These lines contain #> invalid UTF-8 characters: 134
txt[134]
#> [1] "Error: package or namespace load failed for ‘stats\xe2\x80During startup - Warning messages:"
```

Reply:
Do you think you can share that file here?

Reply:
> From my understanding, `useBytes = TRUE` is generally a good idea since it prevents R from translating to native encoding, 

That link is about _writing_ functions. For _reading_ functions it means something different, see the manual of `readChar()`, but also read on for a twist.

> What I do know for sure is that useBytes = TRUE fixes the 1st error in #150, 

That is somewhat surprising and also sad, because according to the `readChar()` docs it should only specify how the `nchars` argument is interpreted, and since rcmdcheck specifies the number of bytes there already, it should not matter at all. (It is fine to overestimate the number of bytes.) So that's a bug in R, most likely in the docs:

```r
❯ writeBin(as.raw(252), tmp <- tempfile())
❯ readChar(tmp, nchars = 1)
Error in readChar(tmp, nchars = 1) : invalid UTF-8 input in readChar()
❯ readChar(tmp, nchars = 1, useBytes = TRUE)
[1] "\xfc"
```

Because, apparently, without `useBytes = TRUE` you cannot read a file that is not in the native encoding. Clearly `useBytes` is not only about `nchars`.

> Also, interestingly, I can read in the Rout.fail file with brio::read_file(), but it still errors out on nchar()

brio does not check if the file is UTF-8, it just _assumes_ it and also tells R that it is, and then R's `nchar()` fails on it because it is not.

> Perhaps the 'correct' fix here is for the Rout.fail files to be encoded as UTF-8 when they're written to disk

No, that is not (easily) possible currently. Base R functions always write to the standard output/error in the native encoding. So it will be UTF-8 on some systems and not UTF-8 on others.

Reply:
> Do you think you can share that file here?

https://www.dropbox.com/s/f7mgxu13po54hcr/ipc-testthat.Rout.fail?dl=0